### PR TITLE
Tramstation Tram "Upgrades"

### PR DESCRIPTION
**Pull request by MMMiracles**
## About The Pull Request
Gives the tram some extra destruction capability while fixing known unintentional tram breaks. Also moves the perma mouse out of the one room with the exposed APC for the entire perma wing.

## Why It's Good For The Game
Tram breaks the crew, not crew break tram.

## Changelog
:cl: MMMiracles
add: Tramstation's Tram is now more capable of making things regret being in the way of it
fix: Tramstation's Tram is now less likely to unintentionally explode
/:cl:

- The tram now smashes through walls instead of phasing through them. Spawns sparks and shakes the screen for nearby observers as the tram kool-aids its way through obstructions.
- The tram now equal-opportunity smashes through most machinery and structures. If a structure is normally not anchored by default (girders, lockers, closets, ect), it will unanchor them and throw them while doing some damage. Machinery now actually gets destroyed when hit by the tram, which also fixes the issue with the recycler breaking the tram.
- Tram platforms themselves have been made inherently indestructible to prevent them from ever breaking in any unexpected gameplay method. 
- The permabrig mouse has been moved to the general activity room instead of the room with the exposed APC wiring.

### CODE REVIEW

File: code/game/objects/structures/industrial_lift.dm [hunk 1]
```diff
@@ -227,16 +230,36 @@ GLOBAL_LIST_EMPTY(lifts)
...
 	if(going == DOWN)
 		for(var/mob/living/crushed in destination.contents)
 			to_chat(crushed, "<span class='userdanger'>You are crushed by [src]!</span>")
 			crushed.gib(FALSE,FALSE,FALSE)//the nicest kind of gibbing, keeping everything intact.
 	else if(going != UP) //can't really crush something upwards
-		for(var/obj/structure/anchortrouble in destination.contents)
-			if(!QDELETED(anchortrouble) && anchortrouble.anchored && (!istype(anchortrouble, /obj/structure/holosign)) && anchortrouble.layer >= GAS_PUMP_LAYER) //to avoid pipes, wires, etc
+		var/atom/throw_target = get_edge_target_turf(src, turn(going, pick(45, -45))) //finds a spot to throw the victim at for daring to be hit by a tram
+		for(var/obj/structure/victimstructure in destination.contents)
+			if(!QDELETED(victimstructure) && (!istype(victimstructure, /obj/structure/holosign)) && victimstructure.layer >= GAS_PUMP_LAYER) //to avoid wires
```
Comment (Kylerace) [inline code/game/objects/structures/industrial_lift.dm:248]: Is there a check for incorporeal structures you can do that isn't an istype check?

File: code/game/objects/structures/industrial_lift.dm [hunk 2]
```diff
@@ -227,16 +230,36 @@ GLOBAL_LIST_EMPTY(lifts)
...
+					victimstructure.anchored = FALSE
+					victimstructure.take_damage(rand(20,25))
+					victimstructure.throw_at(throw_target, 200, 4)
+		for(var/obj/machinery/victimmachine in destination.contents)
+			if(!QDELETED(victimmachine) && victimmachine.layer >= GAS_PUMP_LAYER) //to avoid pipes
 				playsound(src, 'sound/effects/bang.ogg', 50, TRUE)
-				visible_message("<span class='notice'>[src] smashes through [anchortrouble]!</span>")
-				anchortrouble.deconstruct(FALSE)
+				visible_message("<span class='danger'>[src] smashes through [victimmachine]!</span>")
+				victimmachine.Destroy()
```
Comment (Kylerace) [inline code/game/objects/structures/industrial_lift.dm:262]: Why Destroy()? You mean qdel() or deconstruct()?

Comment (MMMiracles) [reply code/game/objects/structures/industrial_lift.dm:262]: Destroy doesn't leave a frame like deconstruct does, and it seems like deconstruct checks to see if it doesn't have the NODECONSTRUCT_1 flag.

Comment (Kylerace) [reply code/game/objects/structures/industrial_lift.dm:262]: Destroy() is only called by qdel() itself no exceptions. just use qdel

Comment (Fikou) [reply code/game/objects/structures/industrial_lift.dm:262]: use qdel

**Comment by Farquaar** [comment]
Tram powercreep

**Comment by Zytolg** [comment]
> Tram powercreep

I hope so. Tram Conductor is a real job that you can get from the HoP after all!

**Comment by moocowswag** [comment]
Does tram hit mime's invisible wall and other special invincible walls like wiz's force wall?

**Review by Maurukas — APPROVED** [review]

File: code/game/objects/structures/industrial_lift.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/game/objects/structures/industrial_lift.dm:248]: Seconding kyler's question, also you should continue if the object is QDELETED, rather then checking to see if it's not. See https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-early-return

It just makes things easier to read

File: code/game/objects/structures/industrial_lift.dm [hunk 3]
```diff
@@ -227,16 +230,36 @@ GLOBAL_LIST_EMPTY(lifts)
...
+				if(victimstructure.anchored && initial(victimstructure.anchored) == TRUE) //prevents you from wrenching lockers/crates to easily break them open
+					visible_message("<span class='danger'>[src] smashes through [victimstructure]!</span>")
+					victimstructure.deconstruct(FALSE)
+				else
+					visible_message("<span class='danger'>[src] violently rams [victimstructure] out of the way!</span>") 
+					victimstructure.anchored = FALSE
+					victimstructure.take_damage(rand(20,25))
+					victimstructure.throw_at(throw_target, 200, 4)
+		for(var/obj/machinery/victimmachine in destination.contents)
+			if(!QDELETED(victimmachine) && victimmachine.layer >= GAS_PUMP_LAYER) //to avoid pipes
```
Comment (LemonInTheDark) [inline code/game/objects/structures/industrial_lift.dm:259]: Ditto here

File: code/game/objects/structures/industrial_lift.dm [hunk 4]
```diff
@@ -227,16 +230,36 @@ GLOBAL_LIST_EMPTY(lifts)
...
 			to_chat(crushed, "<span class='userdanger'>You are crushed by [src]!</span>")
 			crushed.gib(FALSE,FALSE,FALSE)//the nicest kind of gibbing, keeping everything intact.
 	else if(going != UP) //can't really crush something upwards
-		for(var/obj/structure/anchortrouble in destination.contents)
-			if(!QDELETED(anchortrouble) && anchortrouble.anchored && (!istype(anchortrouble, /obj/structure/holosign)) && anchortrouble.layer >= GAS_PUMP_LAYER) //to avoid pipes, wires, etc
+		var/atom/throw_target = get_edge_target_turf(src, turn(going, pick(45, -45))) //finds a spot to throw the victim at for daring to be hit by a tram
+		for(var/obj/structure/victimstructure in destination.contents)
+			if(!QDELETED(victimstructure) && (!istype(victimstructure, /obj/structure/holosign)) && victimstructure.layer >= GAS_PUMP_LAYER) //to avoid wires
+				playsound(src, 'sound/effects/bang.ogg', 50, TRUE)
+				if(victimstructure.anchored && initial(victimstructure.anchored) == TRUE) //prevents you from wrenching lockers/crates to easily break them open
```
Comment (LemonInTheDark) [inline code/game/objects/structures/industrial_lift.dm:250]: Is the initial value of anchored really the best thing to be checking here? Don't some lockers  start out as anchored?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/game/objects/structures/industrial_lift.dm (hunk 4 above)
Comment (MMMiracles) [reply code/game/objects/structures/industrial_lift.dm:250]: Is there a better way to check against this? My main goal here is preventing people from anchoring lockers/crates to the track and as far as I'm aware, none of the lockers/crates that spawn on-station or can be ordered are anchored by default. Adding a typecheck specifically for lockers/crates seem a bit much and I'm not aware of any that start anchored.

File: code/game/objects/structures/industrial_lift.dm (hunk 1 above)
Comment (MMMiracles) [reply code/game/objects/structures/industrial_lift.dm]: As far as I'm aware there isn't any real 'check' for incorporeal structures past something like making a new isincorporeal proc or something weird like checking to see if it's initially dense or w/e

File: code/game/objects/structures/industrial_lift.dm (hunk 4 above)
Comment (Dennok) [reply code/game/objects/structures/industrial_lift.dm:249]: Just unanchor it and push with no damage, if it contayner.

Comment (LemonInTheDark) [reply code/game/objects/structures/industrial_lift.dm:249]: > 
> 
> Is there a better way to check against this? My main goal here is preventing people from anchoring lockers/crates to the track and as far as I'm aware, none of the lockers/crates that spawn on-station or can be ordered are anchored by default. Adding a typecheck specifically for lockers/crates seem a bit much and I'm not aware of any that start anchored.

Sorry for leaving this so long, I guess things will become clear if this is a proper issue, smells a bit but what can you do

File: code/game/objects/structures/industrial_lift.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/game/objects/structures/industrial_lift.dm]: pain

**Comment by LemonInTheDark** [comment]
Fix the conflicts and it's a go

**Pull request merged**